addons/mail/models/discuss/ir_websocket.py

```diff
@@ -82,25 +88,81 @@ def _build_bus_channel_list(self, channels):
         channels.extend([*all_user_channels, *member_specific_channels])
         return super()._build_bus_channel_list(channels)
 
+    def _prepare_subscribe_data(self, channels, last):
+        data = super()._prepare_subscribe_data(channels, last)
+        str_presence_channels = {
+            c for c in channels if isinstance(c, str) and c.startswith("odoo-presence-")
+        }
+        presence_channels = self._build_presence_channel_list(
+            [tuple(c.replace("odoo-presence-", "").split("_")) for c in str_presence_channels]
+        )
+        # There is a gap between a subscription client side (which is debounced)
+        # and the actual subcription thus presences can be missed. Send a
+        # notification to avoid missing presences during a subscription.
+        partners = self.env["res.partner"].browse(
+            [p.id for p, _ in presence_channels if isinstance(p, self.pool["res.partner"])]
+        )
+        guest_ids = [g.id for g, _ in presence_channels if isinstance(g, self.pool["mail.guest"])]
+        # sudo: res.partner - can acess users of partner channels to find
+        # their presences as those channels were already verified during
+        # `_build_bus_channel_list`.
+        domain = expression.AND(
+            [
+                [("last_poll", ">", datetime.now() - timedelta(seconds=2))],
+                expression.OR(
+                    [
+                        [("user_id", "in", partners.with_context(active_test=False).sudo().user_ids.ids)],
+                        [("guest_id", "in", guest_ids)]
+                    ]
+                ),
+            ]
+        )
+        # sudo: mail.presence: can access presences linked to presence channels.
+        missed_presences = self.env["mail.presence"].sudo().search(domain)
+        all_channels = OrderedSet(presence_channels)
+        all_channels.update(
+            self._build_bus_channel_list([c for c in channels if c not in str_presence_channels])
+        )
+        data.update({"channels": all_channels, "missed_presences": missed_presences})
+        return data
+
+    def _subscribe(self, og_data):
+        super()._subscribe(og_data)
+        data = self._prepare_subscribe_data(og_data["channels"], og_data["last"])
+        if bus_target := self._get_missed_presences_bus_target():
+            data["missed_presences"]._send_presence(bus_target=bus_target)
+
     @add_guest_to_context
     def _update_bus_presence(self, inactivity_period, im_status_ids_by_model):
         super()._update_bus_presence(inactivity_period, im_status_ids_by_model)
-        if not self.env.user or self.env.user._is_public():
+        if self.env.user and not self.env.user._is_public():
```

Comment:
Those two branching could be unified

```py
        guest = self.env["mail.guest"]._get_guest_from_context()
        presence = self.env["mail.presence"].sudo() if guest else self.env["mail.presence"]
        if guest or (self.env.user and not self.env.user._is_public()):
            presence.update_presence(
                inactivity_period,
                identity_field="guest_id" if guest else "user_id",
                identity_value=guest.id if guest else self.env.user.id,
            )
 ```

Reply:
Isn't it guaranteed a guest or a user will exist? Thus removing the if?

Reply:
I'm not 100% sure. I believe this scenario occurred for live chats before we refactor the module to use guests. If it works in live chat and the tests are passing then it can probably be removed.